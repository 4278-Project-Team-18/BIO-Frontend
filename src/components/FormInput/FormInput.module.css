.form-input-container {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.form-input-inner-container {
  width: 90%;
}

.form-input-standard {
  width: calc(100% - 24px);
  padding: 4px 12px;
  margin-bottom: 15px;
  vertical-align: middle;
  border-radius: 10px;
  min-height: 36px;
  background-color: #ffffff;
  border: 1px solid rgba(36, 28, 21, 0.3);
  transition: all 0.2s ease-in-out 0s;
  font-size: 16px;
  line-height: 18px;
  font-weight: normal;
  font-family: "Poppins", sans-serif !important;
  letter-spacing: 1px;
}

.form-input-compact {
  width: calc(100% - 24px);
  padding: 0px 12px;
  margin-bottom: 10px;
  vertical-align: middle;
  border-radius: 10px;
  min-height: 36px;
  background-color: #ffffff;
  border: 1px solid rgba(36, 28, 21, 0.3);
  transition: all 0.2s ease-in-out 0s;
  font-size: 14px;
  line-height: 16px;
  font-weight: normal;
  font-family: "Poppins", sans-serif !important;
  letter-spacing: 1px;
}

.form-input-large {
  width: calc(100% - 24px);
  padding: 4px 12px;
  margin-bottom: 15px;
  vertical-align: middle;
  border-radius: 10px;
  min-height: 36px;
  background-color: #ffffff;
  border: 1px solid rgba(36, 28, 21, 0.3);
  transition: all 0.2s ease-in-out 0s;
  font-size: 18px;
  line-height: 20px;
  font-weight: normal;
  font-family: "Poppins", sans-serif !important;
  letter-spacing: 1px;
}

.form-input-standard:focus {
  outline: none;
  border: 1px solid #37b4d0;
  box-shadow: inset 0 0 0 1px #37b4d0;
}

.form-input-compact:focus {
  outline: none;
  border: 1px solid #37b4d0;
  box-shadow: inset 0 0 0 1px #37b4d0;
}

.form-input-label-no-opt-standard {
  font-size: 12px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  margin-bottom: 5px;
  margin-left: 3px;
  color: #aeaeae;
}

.form-input-label-opt-standard {
  font-size: 12px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  margin-bottom: 5px;
  margin-left: 3px;
  color: #aeaeae;
  display: flex;
  flex-direction: column;
}

.form-input-label-no-opt-compact {
  font-size: 10px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  margin-bottom: 2px;
  margin-left: 3px;
  color: #aeaeae;
}

.form-input-label-opt-compact {
  font-size: 10px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  margin-bottom: 0px;
  margin-left: 3px;
  color: #aeaeae;
  display: flex;
  flex-direction: column;
}

.form-input-error {
  font-size: 10px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  margin-bottom: 5px;
  margin-left: 3px;
  color: #fe4d4d;
}

.form-input-opt-text {
  font-size: 10px;
  font-weight: 300;
  letter-spacing: -1px;
  color: #c4c4c4;
}
