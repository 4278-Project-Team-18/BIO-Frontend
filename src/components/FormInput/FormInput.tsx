import styles from "./FormInput.module.css";
import {
  FormInputSizeVariant,
  type FormInputProps,
} from "./FormInput.definitions";
import { Controller } from "react-hook-form";

/**
 * FormInput component
 *
 * @param {string} name - name of the input (for react-hook-form data)
 * @param {string} type - type of the input (e.g. text, password, email)
 * @param {string} label - label of the input (optional, appears top left)
 * @param {string} placeholder - placeholder of the input (optional)
 * @param {string} defaultValue - default value of the input (optional)
 * @param {string} optText - optional text of the input (optional, displayed in gray next to label in parentheses)
 * @param {any} control - control of the input (from react-hook-form)
 * @param {string} error - error of the input (optional, displayed in red next to label)
 */
const FormInput = ({
  name,
  type,
  label,
  placeholder,
  defaultValue,
  optText,
  control,
  error,
  autocomplete,
  sizeVariant = FormInputSizeVariant.standard,
  extraStyles = {},
  paragraph = false,
}: FormInputProps) => (
  <Controller
    name={name}
    control={control}
    render={({ field }) => (
      <div className={styles[`form-input-container`]}>
        <div className={styles["form-input-inner-container"]}>
          {(label || error) && (
            <div
              className={
                error
                  ? styles[`form-input-label-no-opt-${sizeVariant}`]
                  : styles[`form-input-label-opt-${sizeVariant}`]
              }
            >
              <div className={styles["label"]}>{label}</div>
              {!error && optText && (
                <span className={styles["form-input-opt-text"]}>
                  {" (" + optText + ")"}
                </span>
              )}
              {error && (
                <span className={styles["form-input-error"]}>
                  {" (" + error + ")"}
                </span>
              )}
            </div>
          )}
          {!paragraph ? (
            <input
              onChange={field.onChange}
              onBlur={field.onBlur}
              name={field.name}
              ref={field.ref}
              className={styles[`form-input-${sizeVariant}`]}
              placeholder={placeholder}
              type={type}
              defaultValue={defaultValue}
              autoComplete={autocomplete}
              style={{ ...extraStyles }}
            />
          ) : (
            <textarea
              onChange={field.onChange}
              onBlur={field.onBlur}
              name={field.name}
              ref={field.ref}
              className={`${styles[`form-input-${sizeVariant}`]} ${
                styles["form-textarea"]
              }`}
              placeholder={placeholder}
              defaultValue={defaultValue}
              autoComplete={autocomplete}
              style={{ ...extraStyles }}
            />
          )}
        </div>
      </div>
    )}
  />
);

export default FormInput;
