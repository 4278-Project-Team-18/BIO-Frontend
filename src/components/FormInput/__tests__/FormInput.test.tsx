import FormInput from "../FormInput";
import { render, renderHook } from "@testing-library/react";
import { useForm } from "react-hook-form";

describe("FormInput Tests", () => {
  test("FormInput should render correctly", () => {
    // Create a mock control
    const { result } = renderHook(() => useForm());
    const control = result.current.control;

    // Create mock props
    const mockProps = {
      name: "test",
      type: "text",
      label: "Test Label",
      placeholder: "Test Placeholder",
      defaultValue: "Test Value",
      optText: "Test OptText",
      control: control,
      error: "Test Error",
    };

    const component = render(<FormInput {...mockProps} />);

    // Check if the component is rendered
    expect(component).toBeTruthy();

    // Check if the label is correct
    expect(component.getByText(mockProps.label)).toBeTruthy();

    // Check if the placeholder is correct
    expect(component.getByPlaceholderText(mockProps.placeholder)).toBeTruthy();

    // Check if the value is correct
    expect(component.getByDisplayValue(mockProps.defaultValue)).toBeTruthy();
  });

  test("FormInput snapshot should render correctly", () => {
    // Create a mock control
    const { result } = renderHook(() => useForm());
    const control = result.current.control;

    // Create mock props
    const mockProps = {
      name: "test",
      type: "text",
      label: "Test Label",
      placeholder: "Test Placeholder",
      defaultValue: "Test Value",
      optText: "Test OptText",
      control: control,
      error: "Test Error",
    };

    const component = render(<FormInput {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
