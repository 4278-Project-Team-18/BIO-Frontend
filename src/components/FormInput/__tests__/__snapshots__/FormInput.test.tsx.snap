// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput Tests FormInput snapshot should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <div
        class="form-input-container"
      >
        <div
          class="form-input-inner-container"
        >
          <div
            class="form-input-label-no-opt"
          >
            Test Label
            <span
              class="form-input-error"
            >
               (Test Error)
            </span>
          </div>
          <input
            class="form-input"
            name="test"
            placeholder="Test Placeholder"
            type="text"
            value="Test Value"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-input-container"
    >
      <div
        class="form-input-inner-container"
      >
        <div
          class="form-input-label-no-opt"
        >
          Test Label
          <span
            class="form-input-error"
          >
             (Test Error)
          </span>
        </div>
        <input
          class="form-input"
          name="test"
          placeholder="Test Placeholder"
          type="text"
          value="Test Value"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
